File: client/src/api/secure_storage.ts
/**
 * OS keychain 包装：Win Credential Manager / macOS Keychain / Linux Secret Service。
 * 比 localStorage 安全，token 不会被网页注入读取。
 */
import { invoke } from '@tauri-apps/api/core'

export const SECURE_KEYS = {
  ACCESS_TOKEN: 'access_token',
  REFRESH_TOKEN: 'refresh_token',
  USER_ID: 'user_id',
  BYOK_CONFIG: 'byok_config',  // M3: BYOK provider 配置（API key + model + base_url）JSON-encoded
} as const

export type SecureKey = (typeof SECURE_KEYS)[keyof typeof SECURE_KEYS]

export async function secureSet(key: SecureKey, value: string): Promise<void> {
  await invoke('secure_set', { key, value })
}

export async function secureGet(key: SecureKey): Promise<string | null> {
  const v = await invoke<string | null>('secure_get', { key })
  return v ?? null
}

export async function secureDelete(key: SecureKey): Promise<void> {
  await invoke('secure_delete', { key })
}

File: client/src-tauri/src/commands/secure_storage.rs
// OS keychain 包装：Win Credential Manager / macOS Keychain / Linux Secret Service。
// 用于持有 access_token / refresh_token / BYOK API keys。
use keyring::Entry;
use serde::Serialize;

const SERVICE: &str = "scholarpilot";

#[derive(Debug, Serialize)]
pub struct SecureStorageError {
    message: String,
}

impl From<keyring::Error> for SecureStorageError {
    fn from(err: keyring::Error) -> Self {
        Self { message: err.to_string() }
    }
}

#[tauri::command]
pub fn secure_set(key: String, value: String) -> Result<(), SecureStorageError> {
    let entry = Entry::new(SERVICE, &key)?;
    entry.set_password(&value)?;
    Ok(())
}

#[tauri::command]
pub fn secure_get(key: String) -> Result<Option<String>, SecureStorageError> {
    let entry = Entry::new(SERVICE, &key)?;
    match entry.get_password() {
        Ok(v) => Ok(Some(v)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(e.into()),
    }
}

#[tauri::command]
pub fn secure_delete(key: String) -> Result<(), SecureStorageError> {
    let entry = Entry::new(SERVICE, &key)?;
    match entry.delete_password() {
        Ok(_) => Ok(()),
        Err(keyring::Error::NoEntry) => Ok(()), // 已不存在视为成功
        Err(e) => Err(e.into()),
    }
}
